reps/2024-12-4-ray-service-incr-upgrade.md

```diff
@@ -0,0 +1,125 @@
+
+max_surge_percent  (int [0, 100])
+- max_surge_percent represents the maximum allowed percentage increase in capacity during a scaling operation. It limits how much the new total capacity can exceed the original target capacity. 
+    - The formula is: target_capacity_old + target_capacity_new <= (100% + max_surge_percent)
+
+After adding the above field to the Ray Serve schema, these APIs will be added to the KubeRay CR spec and can be specified by the user as follows:
+
+```sh
+type RayServiceUpgradeSpec struct {
+  // Type represents the strategy used when upgrading the RayService. Currently supports `NewCluster` and `None`.
+  Type *RayServiceUpgradeStrategy `json:"type,omitempty"`
+  // The target capacity percentage of the upgraded RayCluster.
+  // Defaults to 100% target capacity.
+  // +kubebuilder:default:=100
+  TargetCapacity *int32 `json:"targetCapacity,omitempty"`
+  // The maximum percent increase of the total capacity during a scaling operation.
+  // Defaults to 100%, i.e. a new RayCluster with equal capacity can be scaled up.
+  // +kubebuilder:default:=100
+  MaxSurgePercent *int32 `json:"maxSurgePercent,omitempty"`
+}
+```
+
+The RayService spec would then look as follows:
+```sh
+// RayServiceSpec defines the desired state of RayService
+type RayServiceSpec struct {
+  // UpgradeSpec defines the scaling policy used when upgrading the RayService.
+  UpgradeSpec *RayServiceUpgradeSpec `json:"upgradeSpec,omitempty"`
+  // Defines the applications and deployments to deploy, should be a YAML multi-line scalar string.
+  ServeConfigV2  string         `json:"serveConfigV2,omitempty"`
+  RayClusterSpec RayClusterSpec `json:"rayClusterConfig,omitempty"`
+  ...
+}
+```
+
+Example RayService manifest:
+```sh
+apiVersion: ray.io/v1
+kind: RayService
+metadata:
+  name: example-rayservice
+spec:
+  upgradeStrategy: "NewCluster"
+  upgradeSpec:
+    type: "NewCluster"
+    targetCapacity: 50
+    maxSurgePercent: 50
+  serveConfigV2: |
+    ...
+```
+
+#### Key requirements:
+- If users want to use the original blue/green deployment, RayService should work regardless of whether the K8s cluster has Gateway-related CRDs
+- If users specify both `target_capacity` and `max_surge_percent` Ray should validate whether there is a conflict in the values before proceeding
+
+### Gateway API
+
+Kubernetes Gateway API is a set of resources designed to manage HTTP(S), TCP, and other network traffic for Kubernetes clusters. It provides a more flexible and extensible alternative to the traditional Ingress API, offering better support for service mesh integrations, routing policies, and multi-cluster configurations. Gateway API already has a well-defined interface for traffic splitting, so KubeRay would not need to implement backends for different Ingress controllers.
```

Comment:
which persona is responsible for which Gateway resources? For example, `GatewayClass` is created / maintained by the K8s cluster admin.